snapcraft/plugins/_python/_pip.py

```diff
@@ -0,0 +1,339 @@
+
+    if requirements:
+        for requirement in requirements:
+            args.extend(['--requirement', requirement])
+
+    if process_dependency_links:
+        args.append('--process-dependency-links')
+
+
+class Pip:
+    """Wrapper for pip abstracting the args necessary for use in a part.
+
+    This class takes care of fetching pip, setuptools, and wheel, and then
+    simply shells out to pip with the magical arguments necessary to install
+    packages into a part.
+
+    Of particular importance: packages must be downloaded (via download())
+    before they can be installed or have wheels built.
+    """
+
+    def __init__(self, *, python_major_version, part_dir, install_dir,
+                 stage_dir):
+        """Initialize pip.
+
+        Check to see if pip has already been installed. If not, fetch pip,
+        setuptools, and wheel, and install them so they can be used.
+
+        :param str python_major_version: The python major version to find (2 or
+                                         3)
+        :param str part_dir: Path to the part's working area
+        :param str install_dir: Path to the part's install area
+        :param str stage_dir: Path to the staging area
+
+        :raises MissingPythonCommandError: If no python could be found in the
+                                           staging or part's install area.
+        """
+        self._python_major_version = python_major_version
+        self._part_dir = part_dir
+        self._install_dir = install_dir
+        self._stage_dir = stage_dir
+
+        self._python_package_dir = os.path.join(
+            self._part_dir, 'python-packages')
+        os.makedirs(self._python_package_dir, exist_ok=True)
+
+        self._python_command = get_python_command(
+            self._python_major_version, stage_dir=self._stage_dir,
+            install_dir=self._install_dir)
+        self._python_home = get_python_home(
+            self._python_major_version, stage_dir=self._stage_dir,
+            install_dir=self._install_dir)
+
+        self._setup()
+
+    def _setup(self):
+        # Check to see if we have our own pip, yet. If not, we need to use the
+        # pip on the host (installed via build-packages) to grab our own.
+        if not self._is_pip_installed():
+            logger.info('Fetching pip, setuptools, and wheel...')
+            real_python_home = self._python_home
+
+            # Make sure we're using pip from the host
+            self._python_home = os.path.join(os.path.sep, 'usr')
+
+            # Using the host's pip, install our own pip and other tools we
+            # need.
+            self.download({'pip', 'setuptools', 'wheel'})
+            self.install({'pip', 'setuptools', 'wheel'}, ignore_installed=True)
+
+            # Now that we have our own pip, reset the python home
+            self._python_home = real_python_home
+
+    def _is_pip_installed(self):
+        try:
+            self._run([], stderr=subprocess.STDOUT)
```

Comment:
please add a comment, I had to look at the implementation of `_run` to figure out what this would do.

Reply:
Not a problem, done.